### PR TITLE
Backport of 3.0 Session API hardening and clear().

### PR DESCRIPTION
**Pull request by dereuromark**
Backport of https://github.com/cakephp/cakephp/pull/5543 to allow clearing the whole session without always having to also renew it.

Also added more tests and fixed doc blocks.

### CODE REVIEW

File: lib/Cake/Model/Datasource/CakeSession.php [hunk 1]
```diff
@@ -468,14 +468,19 @@ public static function destroy() {
...
+ * Optionally also clears the session id and renews the session.
+ *
+ * @param bool $renew If the session should also be renewed. Defaults to false.
  * @return void
  */
-	public static function clear() {
+	public static function clear($renew = false) {
 		$_SESSION = null;
-		self::$id = null;
-		self::renew();
```
Comment (markstory) [inline lib/Cake/Model/Datasource/CakeSession.php]: This is an interesting behavior change that may surprise people. The old default was to renew the session, but now its not.

Comment (dereuromark) [inline lib/Cake/Model/Datasource/CakeSession.php]: So for 2.7 better to default to "true" then?
I initially suspected most people to want to just clean, without the renew side effect But keeping BC here might be the wiser thing after all, despite possible migration guide documentation.

Comment (markstory) [inline lib/Cake/Model/Datasource/CakeSession.php]: Part of being backwards compatible is having methods continue to behave the same given the same set of parameters.

**Comment by dereuromark** [comment]
Should be fine now.

**Pull request merged**